src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs

```diff
@@ -13,5 +27,177 @@ public async Task ExecuteMoveDocuments()
+            var removedShards = oldSetting.Shards;
+
+            foreach (var shard in updatedSetting.Shards)
+            {
+                if (oldSetting.Shards.Contains(shard))
+                    removedShards.Remove(shard);
+                else if (configuration.Shards.ContainsKey(shard) == false)
+                    throw new InvalidDataException($"Cannot assign shard number {shard} to prefix {updatedSetting.Prefix}, " +
+                                                   $"there's no shard '{shard}' in sharding topology!");
+            }
+
+            if (removedShards.Count <= 0) 
+                return;
+
+            // check that there are no bucket ranges mapped to these shards
+
+            int index;
+            bool found = false;
+            var prefixBucketRangeStart = oldSetting.BucketRangeStart;
+
+            for (index = 0; index < configuration.BucketRanges.Count; index++)
+            {
+                var range = configuration.BucketRanges[index];
+                if (range.BucketRangeStart < prefixBucketRangeStart)
+                    continue;
+
+                if (range.BucketRangeStart == prefixBucketRangeStart)
+                    found = true;
+
+                break;
+            }
+
+            if (found == false) 
+                return;
+
+            var shards = new List<int>
+            {
+                configuration.BucketRanges[index++].ShardNumber
+            };
+
+            var nextPrefixedRangeStart = prefixBucketRangeStart + ShardHelper.NumberOfBuckets;
+            for (; index < configuration.BucketRanges.Count; index++)
+            {
+                var range = configuration.BucketRanges[index];
+                if (range.BucketRangeStart < nextPrefixedRangeStart)
+                {
+                    shards.Add(range.ShardNumber);
+                    continue;
+                }
+
+                break;
+            }
+
+            foreach (var shard in removedShards)
+            {
+                if (shards.Contains(shard))
+                    throw new InvalidOperationException(
+                        $"Cannot remove shard {shard} from '{updatedSetting.Prefix}' settings in {nameof(ShardingConfiguration)}.{nameof(ShardingConfiguration.Prefixed)}. " +
+                        $"There are bucket ranges mapped to this shard. " +
+                        "In order to remove a shard from a Prefixed setting, first you need to migrate all its buckets to another shard.");
+
+            }
+        }
+
+        private async Task<bool> AssertNoDocumentsStartingWith(TransactionOperationContext context, string prefix, string[] urls, string database = null)
+        {
+            using (var requestExecutor = RequestExecutor.CreateForServer(urls, database ?? DatabaseName, ServerStore.Server.Certificate.Certificate, DocumentConventions.DefaultForServer))
+            {
+                var command = new GetDocumentsCommand(requestExecutor.Conventions, startWith: prefix,
```

Comment:
this doesn't cover the case if the shard has replication factor > 1, but I guess it also part of the docs 
>It is therefore the user's responsibly to make sure no matching docs exist when adding/deleting a prefix.